proj-xcode/Classes/networking/private/client/PA2AsyncOperation.m

```diff
@@ -0,0 +1,174 @@
+/**
+ * Copyright 2018 Wultra s.r.o.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#import "PA2AsyncOperation.h"
+#import "PA2Log.h"
+
+
+@interface PA2AsyncOperation (Private)
+@property (nonatomic, assign) BOOL isExecuting;
+@property (nonatomic, assign) BOOL isFinished;
+@end
+
+
+@implementation PA2AsyncOperation
+{
+	BOOL _reported;
+	BOOL _isExecuting;
+	BOOL _isFinished;
+	BOOL _inExecutionBlock;
+	dispatch_queue_t _reportQueue;
+	id<NSLocking> _lock;
+}
+
+- (id) initWithReportQueue:(dispatch_queue_t)queue
+{
+	self = [super init];
+	if (self) {
+		_reportQueue = queue;
+		_lock = [[NSLock alloc] init];
+	}
+	return self;
+}
+
+
+#pragma mark - main & cancel
+
+- (void) main
```

Comment:
Primary function to override should be `start`. Overriding only `main` will result in default `NSOperation.start` to be called, which is performing some internal state changes (that are overridden and thus don't reflect actual state) like `willChangeValueForKey:@"isExecuting"`